File: backend/app/services/pins.py
"""訊息置頂的共用邏輯:權限、計數、清單。

權限:direct 兩位成員皆可;group 僅 admin(沿用 ConversationMember.role)。
上限:每對話最多 PIN_LIMIT 則。
"""

import uuid

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models import Conversation, Message
from app.services.conversations import get_role_map

PIN_LIMIT = 10


async def can_pin(db: AsyncSession, conv: Conversation, user_id: uuid.UUID) -> bool:
    """是否可釘選/取消(呼叫端須已確認 user 為 conv 成員)。"""
    if conv.type == "direct":
        return True
    roles = await get_role_map(db, conv.id)
    return roles.get(user_id) == "admin"


async def count_pins(db: AsyncSession, conversation_id: uuid.UUID) -> int:
    return (
        await db.execute(
            select(func.count())
            .select_from(Message)
            .where(
                Message.conversation_id == conversation_id,
                Message.pinned_at.is_not(None),
            )
        )
    ).scalar_one()


async def list_pins(db: AsyncSession, conversation_id: uuid.UUID) -> list[Message]:
    """該對話的釘選訊息,pinned_at 由新到舊。"""
    return list(
        (
            await db.execute(
                select(Message)
                .where(
                    Message.conversation_id == conversation_id,
                    Message.pinned_at.is_not(None),
                )
                .order_by(Message.pinned_at.desc())
            )
        ).scalars().all()
    )

File: frontend/chat/src/pins.ts
// 訊息置頂的純邏輯(抽離 React,便於單元測試):釘選權限、釘選列 view model、pins 清單增減。

import type { Conversation, Message } from '../../contracts';

/** 是否可釘選/取消:1對1 雙方皆可;群組僅 admin。 */
export function canPin(conversation: Conversation, userId: string): boolean {
  if (conversation.type === 'direct') return true;
  return conversation.roles[userId] === 'admin';
}

export interface PinnedBarView {
  latest: Message;
  count: number;
}

/** 釘選列顯示資料:最新一則(清單第一筆,後端已 pinned_at desc)+ 總數;空清單回 null。 */
export function pinnedBarView(pins: Message[]): PinnedBarView | null {
  if (pins.length === 0) return null;
  return { latest: pins[0], count: pins.length };
}

/** 新增釘選:置於最前(最新),並依 id 去重。 */
export function addPin(pins: Message[], message: Message): Message[] {
  return [message, ...pins.filter((p) => p.id !== message.id)];
}

/** 移除釘選:濾掉該 id(不存在則原樣回)。 */
export function removePin(pins: Message[], messageId: string): Message[] {
  return pins.filter((p) => p.id !== messageId);
}
